src/fsharp/TypeChecker.fs

```diff
@@ -2059,24 +2059,16 @@ type GeneralizeConstrainedTyparOptions =
 module GeneralizationHelpers = 
     let ComputeUngeneralizableTypars env = 
         
-        // This is just a List.fold. Unfolded here to enable better profiling 
-        let rec loop acc (items: UngeneralizableItem list) =
-             match items with 
-             | [] -> acc
-             | item::rest -> 
-                 let acc = 
-                     if item.WillNeverHaveFreeTypars then 
-                         acc 
-                     else
-                         let ftps = item.GetFreeTyvars().FreeTypars
-                         if ftps.IsEmpty then 
-                             acc 
-                         else 
-                             // These union operations are a performance sore point
-                             unionFreeTypars ftps acc
-                 loop acc rest
+        let acc = Collections.Generic.List()
+        for item in env.eUngeneralizableItems do
+            if not item.WillNeverHaveFreeTypars then
+                let ftps = item.GetFreeTyvars().FreeTypars
+                if not ftps.IsEmpty then
```

Comment:
I think it should stay until we should do a micro perf test on whether it makes a difference.  Allocating an enumerator and stepping it may indeed be more expensive than a non-allocating step.  And emptiness is a very common case.  

We should probably make a struct enumerator for the Zset type to make iteration non-allocating